File: top-interview-questions-medium/dynamic-programming/unique-paths/62.js
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
// var uniquePaths = function(m, n) {
//   if (m === 1 || n === 1) return 1
//   return uniquePaths(m - 1, n) + uniquePaths(m, n - 1)
// }

var uniquePaths = function(m, n) {
  if (m === 1 || n === 1) return 1

  const arr = []

  for (let i = 1; i < m; i++) {
    for (let j = 1; j < n; j++) {
      arr[j] = (arr[j] || 1 ) + (arr[j - 1] || 1)
    }
  }
  return arr[n - 1]
}

File: top-interview-questions-medium/Dynamic Programming/unique-paths/62.py
class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        if m == 1 or n == 1: return 1
        arr = [1] * n
        for i in range(1, m):
            for j in range(1, n):
                arr[j] += arr[j - 1]

        return arr[n - 1]
